recipes/flex/all/conanfile.py

```diff
@@ -0,0 +1,76 @@
+from conans import ConanFile, AutoToolsBuildEnvironment, tools
+from conans.errors import ConanInvalidConfiguration
+import os
+import glob
+
+
+class ConanFileDefault(ConanFile):
+    name = "flex"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/westes/flex"
+    description = "Flex, the fast lexical analyzer generator"
+    topics = ("conan", "flex", "lex", "lexer", "lexical analyzer generator")
+    license = "BSD-2-Clause"    
+
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"shared": False, "fPIC": True}
+
+    requires = ("m4/1.4.18",)
+
+    _source_subfolder = "source_subfolder"
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def configure(self):
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+        if self.settings.os == "Windows":
+            raise ConanInvalidConfiguration("Flex package is not compatible with Windows. Consider using winflexbison instead.")
+
+    def build(self):
+        env_build = AutoToolsBuildEnvironment(self)
+        configure_args = ["--disable-nls", "HELP2MAN=/bin/true", "M4=m4"]
+        if "shared" in self.options and self.options.shared:
+            configure_args.extend(["--enable-shared", "--disable-static"])
+        else:
+            configure_args.extend(["--disable-shared", "--enable-static"])
+
+        with tools.chdir(self._source_subfolder):
+            if self.settings.os == "Linux":
+                # https://github.com/westes/flex/issues/247
+                configure_args.extend(["ac_cv_func_malloc_0_nonnull=yes", "ac_cv_func_realloc_0_nonnull=yes"])
+                # https://github.com/easybuilders/easybuild-easyconfigs/pull/5792
+                configure_args.append("ac_cv_func_reallocarray=no")
+            if tools.cross_building(self.settings, skip_x64_x86=True):
+                # stage1flex must be built on native arch: https://github.com/westes/flex/issues/78
+                self.run("./configure %s" % " ".join(configure_args))
+                env_build.make(args=["-C", "src", "stage1flex"])
+                self.run("mv src/stage1flex src/stage1flex.build")
+                env_build.make(args=["distclean"])
+                with tools.environment_append(env_build.vars):
+                    env_build.configure(args=configure_args)
+                    cpu_count_option = "-j%s" % tools.cpu_count()
+                    self.run("make -C src %s || true" % cpu_count_option)
+                    self.run("mv src/stage1flex.build src/stage1flex")
+                    self.run("touch src/stage1flex")
+                    env_build.make(args=["-C", "src"])
```

Comment:
I would create a `flex-stage1` build requirement recipe.
And use the new conan cross building magic.
Then, the 2 branches for crossbuilding and native building can be merged and `stage1flex` can always be used.

Reply:
That's an interesting idea, but it has the drawback to double maintenance, and double sources in local cache, so I would like to have more opinions about this before spending more time. ping @Croydon @uilianries

Reply:
I just realized that flex 2.6.4 has this commit https://github.com/westes/flex/commit/89f10ee36f95430f18ac70f971fe9f9f894d9dc5, so maybe we don't have to use this workaround in the first place ? Could someone having a realistic cross-compilation setup try to remove this `if tools.cross_building(self.settings, skip_x64_x86=True):` branch and see if it works ?

Reply:
@jgsogo FYI
flex has the same `CC_FOR_BUILD` option as gmp (see https://github.com/conan-io/conan/pull/5594#issuecomment-598265372).
So it would be nice if this is taken into account in future conan cross building modeling.

Reply:
> I just realized that flex 2.6.4 has this commit [westes/flex@89f10ee](https://github.com/westes/flex/commit/89f10ee36f95430f18ac70f971fe9f9f894d9dc5), so maybe we don't have to use this workaround in the first place ? Could someone having a realistic cross-compilation setup try to remove this `if tools.cross_building(self.settings, skip_x64_x86=True):` branch and see if it works ?

This code doesn't work when building for e.g. armv8 because stage1flex will also be built for armv8. (#1413)
What it needs is more cross building support by conan.
But that's a WIP..

Or my stage1flex idea, but that is indeed a duplication of work so a non-starter.